### PR TITLE
Retry-loop for git clone on Travis/Circle

### PR DESCRIPTION
**Pull request by wilzbach**
If we want to merge only if Travis CI passes too (https://github.com/braddr/d-tester/issues/52), then it should be a bit more failure-proof.

This should hopefully better protect us against random network failures like e.g. here https://travis-ci.org/dlang/dmd/jobs/150380398

CC @MartinNowak

### CODE REVIEW

File: circleci.sh [hunk 1]
```diff
@@ -11,6 +11,23 @@ case $CIRCLE_NODE_INDEX in
...
+    path="$2"
+    branch="$3"
+    for i in {0..4}; do
+        if git clone --depth=1 --branch "$branch" "$url" "$path"; then
+            break
+        elif [ $i -lt 4 ]; then
+            sleep $((1 << $i))
+        else
+            echo "Failed to clone: ${url}"
+            exit 1;
```
Comment (MartinNowak) [inline circleci.sh]: redundant semicolon

File: circleci.sh [hunk 2]
```diff
@@ -11,6 +11,23 @@ case $CIRCLE_NODE_INDEX in
     1) MODEL=32 ;;
 esac
 
+# clone druntime and phobos
+clone() {
+    url="$1"
+    path="$2"
+    branch="$3"
```
Comment (MartinNowak) [inline circleci.sh]: use `local` variables

**Comment by MartinNowak** [comment]
There are python scripts for that [git-retry(1)](https://commondatastorage.googleapis.com/chrome-infra-docs/flat/depot_tools/docs/html/git-retry.html), but bash is good enough.

**Comment by MartinNowak** [comment]
Auto-merge toggled on

**Comment by wilzbach** [comment]
> There are python scripts for that git-retry(1), but bash is good enough.

Yes, but this is then the chicken-egg problem of fetching them ;-)

> Auto-merge toggled on

Thanks!

@MartinNowak any other ideas how we could make Travis more stable? The other big issue of failing tests that I saw was failure due to timeouts and unfortunately even with your optimizations, we still run into timeouts e.g. at the [optimization merge](https://travis-ci.org/dlang/dmd/builds/150353083).

image

**Comment by MartinNowak** [comment]
> @MartinNowak any other ideas how we could make Travis more stable? The other big issue of failing tests that I saw was failure due to timeouts and unfortunately even with your optimizations, we still run into timeouts e.g. at the optimization merge.

I guess we should reconsider the self-compilation tests. The main purpose for adding Travis-CI was to test other host compilers, not dmd.
We could do sth. like directly testing gdc, and ldc as host and then run one self-compilation w/ dmd as host compiler.

**Pull request merged**